Missing implementation of some status return values of cvxopt.glpk.ilp()

**Describe the bug**

The function `cvxopt.glpk.ilp()` may return following values as `status`: `'optimal'`, `'feasible'`, `'undefined'`, `'invalid formulation'`, `'infeasible problem'`, `'LP relaxation is primal infeasible'`, `'LP relaxation is dual infeasible'`, `'unknown'`. AFAICS the following values are not handled correctly by CVXPY and cause a KeyError:

* `'invalid formulation'`
* `'infeasible problem'`

Help text of `ilp()` mentions possible return values::

```
>>> from cvxopt.glpk import ilp
>>> help(ilp)
ilp(...)
    Solves a mixed integer linear program using GLPK.
    
    (status, x) = ilp(c, G, h, A, b, I, B)
...
    status       if status is 'optimal', 'feasible', or 'undefined',
                 a value of x is returned and the status string 
                 gives the status of x.  Other possible values of              status are:  'invalid formulation', 
                 'infeasible problem', 'LP relaxation is primal 
                 infeasible', 'LP relaxation is dual infeasible', 
                 'unknown'.
    
    x            a (sub-)optimal solution if status is 'optimal', 
                 'feasible', or 'undefined'.  None otherwise

```

These are the values which can are handled atm AFAICS:

https://github.com/cvxgrp/cvxpy/blob/b5b0f8b15b1f10106301d2b921ccfc6cbbbaf044/cvxpy/reductions/solvers/conic_solvers/cvxopt_conif.py#L53-L63

There seem to be some unnecessary values too, they might be used by other subclasses. I haven't checked other versions of CVXOPT, this might have changed over time of course.


**To Reproduce**
I have a huge a example which can reproduce the problem. Unfortunately I haven't figured out yet how to reproduce it with a minimal problem. I'll add it to the ticket as soon as I get this done.

If one simply uses a problem with contradicting constraints, cvxopt will return `'LP relaxation is primal infeasible'`. I'm not sure what `'infeasible problem'` means exactly and how to construct one.

**Expected behavior**

The problem should be actually infeasible, so I expected `cp.Problem.solve()` to return without raising a `KeyError` exception and with the status field set to `infeasible` or something similar.

**Output**

That's what PyCharm says when I run my unit tests on my project:

```
...
/opt/miniconda3/envs/myenv/lib/python3.7/site-packages/cvxpy/problems/problem.py:395: in solve
    return solve_func(self, *args, **kwargs)
/opt/miniconda3/envs/myenv/lib/python3.7/site-packages/cvxpy/problems/problem.py:746: in _solve
    self, data, warm_start, verbose, kwargs)
/opt/miniconda3/envs/myenv/lib/python3.7/site-packages/cvxpy/reductions/solvers/solving_chain.py:326: in solve_via_data
    solver_opts, problem._solver_cache)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cvxpy.reductions.solvers.conic_solvers.glpk_mi_conif.GLPK_MI object at 0x7fc0ec038510>
data = {'A': None, 'G': <16x12 sparse matrix, tc='d', nnz=44>, 'b': None, 'bool_vars_idx': [0, 4, 1, 5, 2, 6, ...], ...}
warm_start = True, verbose = False, solver_opts = {}, solver_cache = {}

    def solve_via_data(self, data, warm_start, verbose, solver_opts, solver_cache=None):
        import cvxopt
        import cvxopt.solvers
        # Save original cvxopt solver options.
        old_options = cvxopt.glpk.options.copy()
        # Silence cvxopt if verbose is False.
        if verbose:
            cvxopt.glpk.options["msg_lev"] = "GLP_MSG_ON"
        else:
            cvxopt.glpk.options["msg_lev"] = "GLP_MSG_OFF"
    
        # Apply any user-specific options.
        # Rename max_iters to maxiters.
        if "max_iters" in solver_opts:
            solver_opts["maxiters"] = solver_opts["max_iters"]
        for key, value in solver_opts.items():
            cvxopt.glpk.options[key] = value
    
        try:
            results_tup = cvxopt.glpk.ilp(data[s.C],
                                          data[s.G],
                                          data[s.H],
                                          data[s.A],
                                          data[s.B],
                                          set(int(i) for i in data[s.INT_IDX]),
                                          set(int(i) for i in data[s.BOOL_IDX]))
            results_dict = {}
            results_dict["status"] = results_tup[0]
            results_dict["x"] = results_tup[1]
        # Catch exceptions in CVXOPT and convert them to solver errors.
        #except ValueError as e:
        except RuntimeError as e:
            results_dict = {"status": "unknown"}
    
        # Restore original cvxopt solver options.
        self._restore_solver_options(old_options)
    
        # Convert CVXOPT results to solution format.
        solution = {}
>       status = self.STATUS_MAP[results_dict['status']]
E       KeyError: 'infeasible problem'

/opt/miniconda3/envs/myenv/lib/python3.7/site-packages/cvxpy/reductions/solvers/conic_solvers/glpk_mi_conif.py:90: KeyError
```

**Version**
 - OS: Ubuntu
 - CVXPY Version: 

``` >>> cp.__version__                                                                                                                  
'1.1.3'
>>> cvxopt.__version__                                                                                                              
'1.2.5'
```

**Additional context**

This seems to be very similar to  #812, but since I use CVXPY 1.1.3 the fix from #910 should be included already. So I guess this is still open.

--- a/cvxpy/reductions/solvers/conic_solvers/cvxopt_conif.py
+++ b/cvxpy/reductions/solvers/conic_solvers/cvxopt_conif.py
@@ -52,7 +52,7 @@ class CVXOPT(ECOS):
     # Map of CVXOPT status to CVXPY status.
     STATUS_MAP = {"optimal": s.OPTIMAL,
                   "feasible": s.OPTIMAL_INACCURATE,
-                  "infeasible": s.INFEASIBLE,
+                  "infeasible problem": s.INFEASIBLE,
                   "primal infeasible": s.INFEASIBLE,
                   "LP relaxation is primal infeasible": s.INFEASIBLE,
                   "LP relaxation is dual infeasible": s.UNBOUNDED,

--- a/cvxpy/tests/solver_test_helpers.py
+++ b/cvxpy/tests/solver_test_helpers.py
@@ -516,8 +516,24 @@ def mi_lp_2():
     return sth
 
 
-# TODO: Add a test-case for an infeasible integer-linear program
-#  (infeasible only due integrality constraints).
+def mi_lp_3():
+    # infeasible (but relaxable) test case
+    x = cp.Variable(4, boolean=True)
+    from cvxpy.expressions.constants import Constant
+    objective = cp.Maximize(Constant(1))
+    constraints = [x[0] + x[1] + x[2] + x[3] <= 2,
+                   x[0] + x[1] + x[2] + x[3] >= 2,
+                   x[0] + x[1] <= 1,
+                   x[0] + x[2] <= 1,
+                   x[0] + x[3] <= 1,
+                   x[2] + x[3] <= 1,
+                   x[1] + x[3] <= 1,
+                   x[1] + x[2] <= 1]
+    obj_pair = (objective, -np.inf)
+    con_pairs = [(c, None) for c in constraints]
+    var_pairs = [(x, None)]
+    sth = SolverTestHelper(obj_pair, var_pairs, con_pairs)
+    return sth
 
 
 def mi_socp_1():
@@ -635,6 +651,13 @@ class StandardTestLPs(object):
     @staticmethod
     def test_mi_lp_2(solver, places=4, **kwargs):
         sth = mi_lp_2()
+        sth.solve(solver, **kwargs)
+        sth.verify_objective(places)
+        sth.verify_primal_values(places)
+
+    @staticmethod
+    def test_mi_lp_3(solver, places=4, **kwargs):
+        sth = mi_lp_3()
         sth.solve(solver, **kwargs)
         sth.verify_objective(places)
         sth.verify_primal_values(places)

--- a/cvxpy/tests/test_conic_solvers.py
+++ b/cvxpy/tests/test_conic_solvers.py
@@ -420,6 +420,9 @@ class TestMosek(unittest.TestCase):
     def test_mosek_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='MOSEK')
 
+    def test_mosek_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='MOSEK')
+
     def test_mosek_mi_socp_1(self):
         StandardTestSOCPs.test_mi_socp_1(solver='MOSEK')
 
@@ -615,6 +618,9 @@ class TestCBC(BaseTest):
     def test_cbc_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='CBC')
 
+    def test_cbc_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='CBC')
+
 
 @unittest.skipUnless('GLPK' in INSTALLED_SOLVERS, 'GLPK is not installed.')
 class TestGLPK(unittest.TestCase):
@@ -645,6 +651,9 @@ class TestGLPK(unittest.TestCase):
 
     def test_glpk_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='GLPK_MI')
+
+    def test_glpk_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='GLPK_MI')
 
 
 @unittest.skipUnless('CPLEX' in INSTALLED_SOLVERS, 'CPLEX is not installed.')
@@ -810,6 +819,9 @@ class TestCPLEX(BaseTest):
     def test_cplex_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='CPLEX')
 
+    def test_cplex_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='CPLEX')
+
     def test_cplex_mi_socp_1(self):
         StandardTestSOCPs.test_mi_socp_1(solver='CPLEX', places=3)
 
@@ -943,6 +955,9 @@ class TestGUROBI(BaseTest):
 
     def test_gurobi_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='GUROBI')
+
+    def test_gurobi_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='GUROBI')
 
     def test_gurobi_mi_socp_1(self):
         StandardTestSOCPs.test_mi_socp_1(solver='GUROBI', places=2)
@@ -1087,6 +1102,9 @@ class TestXPRESS(BaseTest):
     def test_xpress_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='XPRESS')
 
+    def test_xpress_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='XPRESS')
+
     def test_xpress_mi_socp_1(self):
         StandardTestSOCPs.test_mi_socp_1(solver='XPRESS')
 
@@ -1172,6 +1190,9 @@ class TestSCIP(unittest.TestCase):
 
     def test_scip_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver="SCIP")
+
+    def test_scip_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver="SCIP")
 
     def get_simple_problem(self):
         """Example problem that can be used within additional tests."""
@@ -1341,6 +1362,9 @@ class TestECOS_BB(unittest.TestCase):
 
     def test_ecos_bb_mi_lp_2(self):
         StandardTestLPs.test_mi_lp_2(solver='ECOS_BB')
+
+    def test_ecos_bb_mi_lp_3(self):
+        StandardTestLPs.test_mi_lp_3(solver='ECOS_BB')
 
     def test_ecos_bb_mi_socp_1(self):
         StandardTestSOCPs.test_mi_socp_1(solver='ECOS_BB')

